[Spring 지하철 노선도 - 1,2단계] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
안녕하세요! 코니😁
디우라고합니다!!ㅎㅎ 만나서 반갑습니다!!

이번 미션을 하며 궁금했던 점은 코드를 처음 받았을 때 `Line` 과 `Station`이 domain 패키지로 구분되어 있더라구요!
그런데 이 둘 모두에서 필드와 Getter만 가지고 있고, DB와 관련된 id를 가지고 있는데, 이런 경우에도 도메인이라고 부를 수 있는건지..궁금합니다..!!

이번 지하철 노선도 미션 잘부탁드립니다!!🙇‍♂️

wishoon committed:
docs: 요구사항 목록 작성

wishoon committed:
feat: 기존에 존재하는 지하철역 이름으로 생성하지 못하게 기능 구현

wishoon committed:
feat: 지하철 노선 등록 구현

wishoon committed:
feat: 지하철 노선 전체 조회 구현

wishoon committed:
feat: 지하철 노선 단건 조회 구현

wishoon committed:
feat: 지하철 노선 수정 구현

wishoon committed:
feat: 지하철 노선 단건 삭제 구현

wishoon committed:
feat: 기존에 존재하는 노선 이름으로 생성하지 못하게 기능 구현

tco0427 committed:
refactor: 불필요한 변수 제거

tco0427 committed:
refactor: 변수명 변경

tco0427 committed:
feat: 데이터베이스 설정

wishoon committed:
test: 지하철역 등록 및 전체 조회 테스트 작성

wishoon committed:
feat: 지하철역 단건 삭제 기능 구현

wishoon committed:
docs: 요구사항 목록 작성

wishoon committed:
refactor: StationDao를 Spring JDBC로 전환

wishoon committed:
refactor: deleteById 로직 성공시 1, 실패시 0을 반환하도록 변경

wishoon committed:
refactor: LineDao를 Spring JDBC로 전환

wishoon committed:
refactor: Station에 대한 Service 계층 도입

wishoon committed:
refactor: StationDao 책임 분리

StationDao에서 INSERT 및 객체를 생성하는 두가지 책임을 분리

wishoon committed:
refactor: Line에 대한 Service 계층 도입

tco0427 committed:
refactor: LineCreateRequest & Response 제거

생성자 오버로딩을 통해서 수정

tco0427 committed:
refactor: 자잘한 수정 진행

불필요한 변수 제거 및 변수명 일관성 있게 변경 등 자잘한 수정을 진행하였습니다.

tco0427 committed:
fix: Line에 대한 deleteById() 테스트 DisplayName

누락된 DisplayName 추가

tco0427 commented:
Service를 도입한 가장 큰 이유는 중복검증을 Service 단에서 하기 위해서 였는데, unique 컬럼에 대해서는 `DuplicateKeyException` 이 던져지더라구요...하지만 도메인과 뷰단의 의존성을 끊어주고, 서비스단에서 DTO를 생성해서 넘겨주기 위해서 Service를 유지하기로 하였습니다.

tco0427 committed:
docs(README): 기능 요구사항 목록 업데이트

choihz commented on src/test/java/wooteco/subway/service/LineServiceTest.java:
```diff
@@ -0,0 +1,86 @@
+import wooteco.subway.dto.LineRequest;
+import wooteco.subway.dto.LineResponse;
+
+@SpringBootTest
+@Transactional
+class LineServiceTest {
```
전체 테스트를 여러 번 돌려 보았는데, 이 테스트 클래스의 테스트들은 실행 순서의 영향을 받는 것으로 보입니다. 가장 먼저 실행될 때는 성공하지만 그렇지 않을 때는 실패하네요.

tco0427 replied:
아직 질문에 대한 내용이 파악이 안되어 DM 드린 상태입니다...!!😅

여러가지 고민해서 우선.. `extracting().contains()` => `extracting.containsExactlyInAnyOrder()` 로 tuple간 순서에 대해 보장하지 않아도 되도록 의심가는 부분을 수정해보았습니다…

choihz commented on src/main/java/wooteco/subway/ui/StationController.java:
```diff
@@ -1,39 +1,46 @@
     public ResponseEntity<StationResponse> createStation(@RequestBody StationRequest stationRequest) {
-        Station station = new Station(stationRequest.getName());
-        Station newStation = StationDao.save(station);
-        StationResponse stationResponse = new StationResponse(newStation.getId(), newStation.getName());
-        return ResponseEntity.created(URI.create("/stations/" + newStation.getId())).body(stationResponse);
+        final StationResponse response = stationService.save(stationRequest);
+
+        return ResponseEntity.created(URI.create("/stations/" + response.getId())).body(response);
     }
```
빈 문자열로도 역이 생성되고 있네요 🤔

tco0427 replied:
지하철역 생성시 공백인지, 그리고 노선을 생성하면서 노선의 이름과 색상이 공백인지를 검증하는 로직을 각 도메인에 추가하였습니다.

각 도메인에 추가한 이유는 이름이 공백인지, 색상이 공백인지에 대한 책임은 각 도메인에 있다고 판단하였기 때문입니다.

또한 IllegalArgumentException을 활용하였습니다. 즉, 공백인 경우 `IllegalArgumentException`을 던지도록 하였는데요, `허용하지 않는 값을 인자로 전달받았을 때` 예외를 발생시키는 `IllegalArgumentException` 이라는 `표준예외`를 활용하여도 충분히 표현이 가능하기 때문입니다. 그리고 이 경우 bad request를 던지도록 하였습니다.

물론 밑에 코멘트 주신 것과 마찬가지로 `IllegalArgumentException이 발생하는 상황이면 모두 BAD REQUEST`가 아닐 수 있기 때문에 앞으로 IllegalArgumentException을 사용할 때 주의가 필요할 것으로 생각됩니다. 혹은 현재는 IllegalArgumentException을 사용하지만 좀 더 구체적인 또 다른 Custom Exception을 만들어 줄 수도 있을 것 같습니다. 하지만 현재는 IllegalArgumentException이 그 의미를 충분히 전달하고, 사용해도 적절하겠다는 판단을 하여 사용하였습니다!😃

choihz commented on src/main/java/wooteco/subway/ui/StationController.java:
```diff
@@ -1,39 +1,46 @@
+        return ResponseEntity.created(URI.create("/stations/" + response.getId())).body(response);
     }
 
     @GetMapping(value = "/stations", produces = MediaType.APPLICATION_JSON_VALUE)
```
기본 제공된 코드이긴 하지만, 혹시 모르고 있었다면 이번 기회에 학습해 보아요. `produces` 설정은 어떤 역할을 할까요? 그리고 지금 이 설정을 지우면 어떻게 될까요?

tco0427 replied:
RequestMapping에서 사용자 요청을 받고, 응답을 보내줄 때, MediaType을 통해서 데이터의 형식을 강제해줄 수 있습니다. `consumes` 의 경우에는 사용자 요청에 의해 들어오는 데이터의 데이터 형식을 정해줄 수 있고, 질문해주신 `produces` 의 경우에는 응답 데이터의 형식을 정해줄 수 있습니다.

언급해주신 코드 부분에서는 "/stations" Get 요청의 응답으로 나가는 List<StationResponse> 가 JSON 형식이도록 강제해준다고 해석할 수 있습니다.
그리고 만약 해당 설정을 지우게 된다면 JSON 형식이 아닌 text/html, XML 등으로도 응답이 가능하게 될 것 같슽니다..!!

또한 @RequestBody에서 HttpMessageConverter를 통해서 해당 json 데이터를 jackson 라이브러리를 통해서 변환해주는 것 같습니다..따라서 지금의 경우 굳이 사용할 필요가 없어보입니다!

choihz commented on src/main/java/wooteco/subway/ui/StationController.java:
```diff
@@ -1,39 +1,46 @@
-        return ResponseEntity.ok().body(stationResponses);
+        final List<StationResponse> response = stationService.findAll();
+
+        return ResponseEntity.ok().body(response);
```
```suggestion
        return ResponseEntity.ok(response);
```

사소한 것이긴 하지만, OK 응답의 경우 이렇게 shortcut을 지원하고 있어요!

tco0427 replied:
`A shortcut for creating a ResponseEntity with the given body and the status set to OK.`와 같이
body()를 생략하는 shortcut을 제공하는지 몰랐네요..!! 감사합니다!!😁

choihz commented on src/main/java/wooteco/subway/ui/LineController.java:
```diff
@@ -0,0 +1,61 @@
+        this.lineService = lineService;
+    }
+
+    @PostMapping("/lines")
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest request) {
+        final Long savedId = lineService.save(request);
+
+        return ResponseEntity.created(URI.create("/lines/" + savedId))
+                .body(new LineResponse(savedId, request.getName(), request.getColor()));
+    }
```
컨트롤러에서 응답 객체를 만드는데, 요청 객체의 값을 바로 꺼내 생성하고 있네요. 이렇게 해도 괜찮을까요?

choihz replied:
또한 응답 형태가 [API 문서](https://techcourse-storage.s3.ap-northeast-2.amazonaws.com/d5c93e187919493da3280be44de0f17f#_%EC%A7%80%ED%95%98%EC%B2%A0_%EB%85%B8%EC%84%A0_%EB%93%B1%EB%A1%9D_http_response)와는 달라요.

```json
{
  "id" : 1,
  "name" : "신분당선",
  "color" : "bg-red-600"
}
```

tco0427 replied:
컨트롤러에서 응답 객체를 만들 때, 요청 객체의 값을 그대로 사용하게 되면, 만약 잘못된 요청에 대한 검증이 제대로 처리되지 않는 경우(ex. 이전에 빈 문자열에 대한 검사가 없는 경우) 그대로 응답에 사용되게 될 것입니다.

또한 만약 Service단에서 예외에 대한 별도의 처리를 하고 그에 대한 값을 반환해준다 라고 할 때, (예를 들어 이름이 ""인 경우 "빈문자열" 이라고 바꿔서 응답해줘야한다.) 그에 대한 처리도 제대로 해주지 못하게 될 것이라고 생각됩니다. (요청값을 그대로 전달하기 때문에 "빈문자열"이 아닌 ""가 응답으로 나가게 될 것이다.)
따라서 lineService.save() 시에 `LineResponse`를 받고, 이를 body에 담아 응답하도록 수정하였습니다.

또한 API 문서와의 스팩이 일치하지 않았던 부분은 `List<StationResponse> stations` 필드를 제거함으로써 일치시켜 주었습니다.😃

choihz commented on src/main/java/wooteco/subway/ui/LineController.java:
```diff
@@ -0,0 +1,61 @@
+        return ResponseEntity.ok().body(responses);
+    }
+
+    @GetMapping("/lines/{id}")
+    public ResponseEntity<LineResponse> showLine(@PathVariable Long id) {
+        final LineResponse response = lineService.findById(id);
+
+        return ResponseEntity.ok().body(response);
+    }
```
존재하지 않는 id로 요청을 보내면 어떤 응답이 오고 있나요?

tco0427 replied:
`EmptyResultDataAccessException` 예외가 발생합니다!! 이 부분에 대한 처리를 추가하도록 하겠습니다😃

choihz commented on src/main/java/wooteco/subway/ui/ExceptionControllerAdvice.java:
```diff
@@ -0,0 +1,15 @@
+@ControllerAdvice
+public class ExceptionControllerAdvice {
+
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> handle(IllegalArgumentException e) {
+        return ResponseEntity.badRequest().body(new ErrorResponse(e.getMessage()));
+    }
```
`IllegalArgumentException`이 발생하는 상황이면 모두 BAD REQUEST인가요? 또한, 이 예외가 아닌 예상치 못한 다른 예외가 발생한다면 어떻게 될까요?

tco0427 replied:
`IllegalArgumentException`이 모두 Bad Request는 아니라고 생각합니다..!! 현재는 중복된 이름인 경우(badRequest)와 존재하지 않는 경우(notFound) 모두 IllegalArgumentException을 던져 bad request로 처리하고 있습니다ㅠ.ㅠ

---

중복된 이름에 대해서는 DAO에서 던져주는 `DuplicateKeyException` 을 사용함으로써 Service에 별도의 검증로직을 두지 않고, 제거해주는 방향으로 개선해보았습니다!
또한 존재하지 않는 경우에 대해서는 NotExistException(custom exception)을 두어서 notFound로 응답할 수 있도록 수정해보았습니다!

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,211 @@
+    @DisplayName("노선을 등록한다.")
+    void createLine() {
+        // given
+        Map<String, String> params = new HashMap<>();
+        params.put("name", "신분당선");
+        params.put("color", "bg-red-600");
```
`Map` 대신 `LineRequest`를 사용하는 건 어떨까요? 어떤 장점이 있을까요?

tco0427 replied:
아하..!! Request DTO를 사용할 생각을 못했던 것 같습니다...!!
지금 변경을 하고 나서 느끼는 장점으로는 테스트코드에서 요청에 대한 관리가 좀 더 편리할 것 같습니다!!

혹시 제가 모르는 또 다른 장점이 있을까요...??ㅠ.ㅠ

choihz replied:
이거 답하려고 생각해 두었는데 지난번에 놓쳤네요!
말씀하신 것처럼 관리가 더 편하죠. 가독성도 높아지고, 또 오타 등으로 실수할 가능성도 확실히 줄어들 거예요. 혹시나 요청 객체의 변경이 이뤄졌다면 테스트 코드에서도 어디가 수정되어야 하는지 빨리 파악할 수 있을 것 같고요.

tco0427 replied:
좋은 내용 감사합니다!!😃

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,211 @@
+                .extract();
+
+        // then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.CREATED.value());
+        assertThat(response.header("Location")).isNotBlank();
```
응답 바디까지 검증해 보는 건 어떨까요?

tco0427 replied:
```java
        assertThat(response.as(LineResponse.class).getName()).isEqualTo("신분당선");
        assertThat(response.as(LineResponse.class).getColor()).isEqualTo("bg-red-600");
```

`as`를 이용한 테스트를 응답 바디를 검증하는 코드를 추가해보았습니다!😃

choihz commented on src/test/java/wooteco/subway/dao/LineDaoTest.java:
```diff
@@ -0,0 +1,99 @@
+import org.springframework.transaction.annotation.Transactional;
+import wooteco.subway.domain.Line;
+
+@SpringBootTest
+@Transactional
+class LineDaoTest {
```
DAO만 테스트하는데 `@SpringBootTest`를 이용해 모든 컨텍스트를 다 띄울 필요가 있을까요? 다른 방법은 없을지 학습해 보아요.

tco0427 replied:
`@SpringBootTest` 대신 JDBC 기반의 컴포넌트들에 초점을 맞춘 `@JdbcTest`를 이용해서 좀 더 가볍게 테스트할 수 있습니다! 또한 `@Transactional` 어노테이션도 포함하고 있어 테스트 이후 롤백에 대한 문제도 없어보입니다!

choihz reviewed (changes requested):
안녕하세요, 디우! 지하철 노선도 미션을 함께할 리뷰어 코니입니다 👋

1, 2단계를 잘 구현해 주셨네요 👍 몇 가지 피드백을 남겨 두었는데요, 보시고 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,52 @@
+package wooteco.subway.domain;
```
> 이번 미션을 하며 궁금했던 점은 코드를 처음 받았을 때 `Line` 과 `Station`이 domain 패키지로 구분되어 있더라구요!
> 그런데 이 둘 모두에서 필드와 Getter만 가지고 있고, DB와 관련된 id를 가지고 있는데, 이런 경우에도 도메인이라고 부를 수 있는건지..궁금합니다..!!

디우가 생각하는 도메인이란 무엇인가요? 😄

tco0427 replied:
저는 도메인이 비즈니스 로직과 관련된 책임을 가지는 객체들이라고 생각합니다. 그에 따른 상태(인스턴스 변수)를 가질 수 있다고 생각합니다. (더크게 보면 내가 해결하고자하는 문제 영역에 해당하는 객체들이라고 볼 수도 있을 것 같고, 그러한 관점이라면 Line과 Station을 domain 패키지로 구분할 수 있을 것 같습니다.)

코드를 처음 받았을 때 Line, Station은 데이터(인스턴스 변수)와 그에 따른 비즈니스 로직을 가지고 있지 않은 상태였으며, id라고 하는 DB와 관련된 인스턴스 변수 또한 가지고 있었습니다...그래서 이것 또한 도메인인가?? 도메인과 Entity의 경계는 무엇일까? 궁금해졌습니다.
제가 아는 Entity는 DB테이블의 컬럼을 필드로 가지는 클래스로 그에 따른 Getter를 가질 수 있는 것으로 이해하고 있습니다. 또한 지금과 같이 식별자(id, Primary Key값)을 가지는게 특징이라고 생각합니다. 그래서 Line과 Station은 도메인보다는 Entity에 가깝지 않나? 하는 생각을 하였습니다!

tco0427 committed:
refactor: contains() -> containsExactlyInAnyOrder()

containsExactlyInAnyOrder() 사용하여 tuple 순서에 영향을 받지 않도록 하였습니다.

tco0427 committed:
refactor: DAO 테스트에서 @JdbcTest 이용

`@JdbcTest`를 사용하여 보다 가볍게 테스트를 수행할 수 있습니다.

tco0427 committed:
refactor: ok응답에 대해서 shortcut 적용

tco0427 committed:
feat: `EmptyResultDataAccessException` 처리

존재하지 않는 id값 등올 조회시 예외가 발생하고, 그에 대한 처리를 해주도록 수정하였습니다.

tco0427 committed:
fix: 의존성 주입 부분 수정

tco0427 committed:
refactor: DAO에서 던져주는 DuplicateKeyException 이용

Service단에서 직접 중복을 검사하는 것이 아니라 unique 컬럼에 의해서 던져지는 `DuplicateKeyException`을 이용하도록 수정하였습니다.

tco0427 committed:
feat: NotExistException 도입

존재하지 않는 Line, Station에 대해서 NotExistException이 던져지도록 Custom Exception을 도입하고 Not Found 를 던지도록 하였습니다.

tco0427 committed:
feat: 지하철역 이름과 노선의 이름 & 색상의 공백 검증

지하철역의 이름, 노선의 이름과 색상이 공백인지를 검증하는 로직을 추가하였습니다.

tco0427 committed:
refactor: `produces` 설정 제거

tco0427 committed:
refactor: 컨트롤러 응답 개선 & API 응답 형태 일치

컨트롤러 응답 부분에서 요청 객체의 값을 그대로 사용하지 않도록 개선하고 API 응답을 문서와 일치하도록 수정하였습니다.

tco0427 committed:
refactor: request DTO를 사용하도록 수정

tco0427 committed:
refactor: 응답바디까지 검증하도록 개선

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,206 @@
+        final LineRequest params = new LineRequest("신분당선", "bg-red-600");
+
+        // when
+        ExtractableResponse<Response> response = RestAssured.given().log().all()
+                .body(params)
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .when()
+                .post("/lines")
+                .then().log().all()
+                .extract();
```
같은 로직이 모든 테스트 메서드에 등장하고 있는데, 별도 메서드로 추출하여 중복을 제거하는 건 어떨까요? 어떤 장점이 있을까요?

tco0427 replied:
RestAssured를 활용하여 API를 테스트하는 부분을 Fixture로 분리하였습니다.

이를 통해서 향후에 또 다른 인수테스트가 추가될 때도 현재 코드를 재활용할 수 있으며,
무엇보다 기존 LineAcceptanceTest나 StationAcceptanceTest의 각 단위테스트 메소드의 흐름을 한 눈에 알 수 있게 가독성이 향상되는게 느껴졌습니다!😃

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,206 @@
+                .extract();
+
+        // then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+        List<Long> expectedLineIds = Stream.of(createResponse1, createResponse2)
+                .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
+                .collect(toList());
+        List<Long> resultLineIds = response.jsonPath().getList(".", LineResponse.class).stream()
+                .map(LineResponse::getId)
+                .collect(toList());
+        assertThat(resultLineIds).containsAll(expectedLineIds);
```
무엇을 검증하고 있는지 눈에 잘 들어오지 않는 것 같아요. 개선해 볼 수 있을까요?

tco0427 replied:
해당 부분은 응답 헤더와 바디로 부터 id값을 꺼내오는 부분입니다.

그런데 `expectedLineIds` 에서도 [API 문서](https://techcourse-storage.s3.ap-northeast-2.amazonaws.com/d5c93e187919493da3280be44de0f17f#Line) 를 보니 응답 바디에서 id 값을 꺼내와도 되겠다는 생각을 할 수 있었습니다.

따라서 두 부분모두 jsonPath()를 활용하여 응답 바디에서 id값을 가져오도록 하는 별도의 메소드를 두어 코드의 가독성을 개선해 보았습니다!

choihz commented on src/test/java/wooteco/subway/service/LineServiceTest.java:
```diff
@@ -0,0 +1,99 @@
+
+@SpringBootTest
+@Transactional
+class LineServiceTest {
```
`-ServiceTest`인데 `@SpringBootTest`로 컨텍스트를 전부 띄워 테스트하고 있어요. 이렇게 했을 때의 장단점에는 어떤 것이 있을까요?

tco0427 replied:
컨텍스트를 띄우기 때문에 DAO에 대한 의존성 주입을 받아서 Service에서 필요한 DAO를 편리하게 주입 받을 수 있다는 장점이 있을 것 같습니다.
하지만 그에 따른 비용이 크다고 생각이 됩니다. 따라서 가짜 DAO객체, 즉 List를 사용하여 메모리에 데이터를 저장하는 방식을 사용하는 DAO를 사용하여 테스트를 진행해보았습니다. 이렇게 되면 컨텍스트를 띄우는데 필요한 시간과 비용을 줄일 수 있고, 원하는 테스트 결과를 얻을 수 있게 됩니다😄

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,57 @@
+
+        final Line savedLine = lineDao.save(line);
+
+        return new LineResponse(savedLine.getId(), savedLine.getName(), savedLine.getColor());
```
프로그램이 계속 커지고 `Line`이 가지고 있는 정보가 늘어난다면, `LineResponse` 역시 수정될 확률이 높을 거예요. `LineResponse`를 생성할 때 이렇게 값을 일일이 꺼내 전달하는 방식을 사용해야 한다면, 수정이 필요할 때 `LineResponse`를 생성하는 모든 부분을 찾아 수정해야 하지 않을까요? 더 나은 방법은 없을까요?

tco0427 replied:
Line 자체를 넘기고 이를 생성자 내부에서 getter를 호출해 필요한 정보를 가져와 초기화를 진행한다면 수정이 필요할 때 생성하는 모든 부분을 찾아서 수정하지않고, 생성자 부분만 수정하면 되므로 좀 더 유연하게 될 것 같습니다.😄

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+    private String name;
+    private String color;
+
+    public Line() {
+    }
```
기본 생성자가 필요한가요?

tco0427 replied:
Line과 Station에서 사용하지 않는 기본생성자를 제거하였습니다😅

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+        this.color = color;
+    }
+
+    private void validateArgument(String name, String color) {
+        if (name.isBlank() || color.isBlank()) {
+            throw new IllegalArgumentException("노선의 이름 혹은 색이 공백일 수 없습니다.");
+        }
+    }
```
지정된 컬럼의 크기를 초과하는 값이 들어오면 어떻게 될까요?

tco0427 replied:
500 응답 코드와 함께 `JdbcSQLDataException`이 발생합니다! 개선해보도록 하겠습니다!

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+package wooteco.subway.domain;
```
> 저는 도메인이 비즈니스 로직과 관련된 책임을 가지는 객체들이라고 생각합니다. 그에 따른 상태(인스턴스 변수)를 가질 수 있다고 생각합니다. (더크게 보면 내가 해결하고자하는 문제 영역에 해당하는 객체들이라고 볼 수도 있을 것 같고, 그러한 관점이라면 Line과 Station을 domain 패키지로 구분할 수 있을 것 같습니다.)
> 
> 코드를 처음 받았을 때 Line, Station은 데이터(인스턴스 변수)와 그에 따른 비즈니스 로직을 가지고 있지 않은 상태였으며, id라고 하는 DB와 관련된 인스턴스 변수 또한 가지고 있었습니다...그래서 이것 또한 도메인인가?? 도메인과 Entity의 경계는 무엇일까? 궁금해졌습니다.
> 제가 아는 Entity는 DB테이블의 컬럼을 필드로 가지는 클래스로 그에 따른 Getter를 가질 수 있는 것으로 이해하고 있습니다. 또한 지금과 같이 식별자(id, Primary Key값)을 가지는게 특징이라고 생각합니다. 그래서 Line과 Station은 도메인보다는 Entity에 가깝지 않나? 하는 생각을 하였습니다!

충분히 잘 이해하고 있으신 걸로 보여요 👍

"도메인과 Entity의 경계는 무엇일까?" 하는 의문이 드셨다고 했는데, 이 경계는 분명한 경우보다 그렇지 않은 경우가 아마 훨씬 많을 거예요. 지금 우리 미션에서도 그런 것처럼요. 앞으로의 미션을 통해서, 그리고 학습을 통해서 배워나가야 할 부분인 것으로 이해해 주셔도 좋을 것 같습니다.

tco0427 replied:
넵!! 감사합니다!!🙏

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+            return false;
+        }
+        Line line = (Line) o;
+        return Objects.equals(name, line.name) && Objects.equals(color, line.color);
```
`name`과 `color`만으로 동등성을 판단한다면 `Line`에서 `id`의 역할은 무엇일까요?

바로 위의 코멘트를 통해 말씀 드린 것처럼, 차근차근 배워나갈 부분이라고 생각하기 때문에 정확한 답을 요구하는 질문은 아니에요. 한번 고민해 보시면 좋을 것 같아 남겨 봅니다. 😊

tco0427 replied:
아직은 id의 역할이 DB의 pk를 자바 애플리케이션 상으로 가져오는 것(ex. id를 통한 조회 등) 이외에는 잘 모르겠습니다....ㅠ.ㅠ 질문에 대해서 조금 더 고민해보도록 하겠습니다...!!

choihz replied:
하나 가정해 보아요. 노선의 이름과 색깔 모두 중복이 가능하도록 비즈니스 로직이 변경되었어요. 그렇다면 이름과 색깔만으로 두 객체가 동등하다고 판단해도 될까요?

tco0427 replied:
만약 이름과 색깔 모두 중복이 가능하게 된다면 PK에 해당하는 id로 두 객체를 판별해야할 것 같아요!!
즉, 요구사항에 따라서 두 객체의 동등성을 비교할 때 id 의 역할이 필요해질 수 있겠다는 생각을 합니다!!

choihz commented on src/main/java/wooteco/subway/ui/ExceptionControllerAdvice.java:
```diff
@@ -0,0 +1,33 @@
+        return ResponseEntity.notFound().build();
+    }
+
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> handle(EmptyResultDataAccessException e) {
+        return ResponseEntity.badRequest().body(new ErrorResponse("해당 값이 존재하지 않습니다."));
+    }
+
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> handle(DuplicateKeyException e) {
+        return ResponseEntity.badRequest().body(new ErrorResponse("이미 존재합니다."));
+    }
```
지금 이 `ExceptionControllerAdvice`가 속한 패키지는 `ui`인데, 이 메서드들이 다루는 예외는 스프링의 `dao` 패키지에 속해 있네요. `dao`의 예외들은 어디서 처리하는 것이 좋을지 고민해 보아도 좋을 것 같아요.

tco0427 replied:
음...해당 질문을 받고 나니, DAO쪽에서 다른 커스텀 예외나 자바 표준예외로 바꾸면 어떨까...하는 생각이 들어요!

아직 JPA에 대해서 학습이 되어 있는 상태는 아니지만! JdbcTemplate에서 JPA를 사용하기로 설계(?)가 변경된다면 지금 현재 사용하는 `package org.springframework.dao` 가 아닌 `javax.persistence.PersistenceException` 의 예외들을 사용하게 될 것 같아요!
따라서 해당 변경의 영향이 Service 계층이나 Controller, 특히 ControllerAdvice 쪽으로 영향을 미치는 것보다는 DAO 내에만 영향을 끼치도록 하는게 좋은 설계라는 생각이 들었습니다!
따라서 DAO쪽에서 예외를 catch하고 다른 커스텀 예외나 자바 표준예외를 활용해보면 어떨까하는 생각이 듭니다!😄

choihz replied:
이미 Service에서 DAO를 사용하고 있는데 이 부분에서의 설계 변경이 일어났을 때 Service에 영향을 미치지 않을 수 있을까요?
괜찮은 방향으로 접근하고 계신 것 같고, 과연 어디서 어떻게 처리하는 것이 더 좋은 방법인지는 다음 단계를 진행하면서 스스로 고민해 보아요 😎 물론 그 과정에서 궁금한 부분이 있으실 때 질문은 언제든 환영입니다.

tco0427 replied:
넵!! 감사합니다~_~

choihz reviewed (approved):
안녕하세요, 디우!

수정해 주신 부분을 확인하고, 추가 피드백을 몇 가지 남겨 보았습니다. 다음 단계를 진행하며 함께 확인 부탁드려요. 그럼, 다음 단계에서 다시 만나요~ 👋
